Web Inspector: Enable Memory profiling in Workers

dcrousso opened the pull request:
#### 4464c381ec4f5eed56465fd656bf37b3835b9a4a
<pre>
Web Inspector: Enable Memory profiling in Workers
<a href="https://bugs.webkit.org/show_bug.cgi?id=167323">https://bugs.webkit.org/show_bug.cgi?id=167323</a>

Reviewed by BJ Burg.

The Web Inspector frontend always assumes the main target when interacting with `Heap`, meaning that even though `Heap` is already enabled inside `Worker` (i.e. there&apos;s even a `WebHeapAgent`) it&apos;ll never communicate with any other `WI.Target`.

Note that there are still a few things missing from full support as `Timelines` currently only supports the main target (e.g. auto-capture in `Worker` on page load won&apos;t happen).

* Source/WebInspectorUI/UserInterface/Models/HeapAllocationsInstrument.js:
(WI.HeapAllocationsInstrument.prototype.startInstrumentation):
(WI.HeapAllocationsInstrument.prototype.stopInstrumentation):
(WI.HeapAllocationsInstrument.prototype._takeHeapSnapshot):
Start and stop `Heap` for each `WI.Target` instead of just the main target.

* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.heapSnapshot):
* Source/WebInspectorUI/UserInterface/Protocol/HeapObserver.js:
(WI.HeapObserver.prototype.trackingStart):
(WI.HeapObserver.prototype.trackingComplete):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotDiffProxy.js:
(WI.HeapSnapshotDiffProxy):
(WI.HeapSnapshotDiffProxy.deserialize):
(WI.HeapSnapshotDiffProxy.prototype.get target): Added.
(WI.HeapSnapshotDiffProxy.prototype.instancesWithClassName):
(WI.HeapSnapshotDiffProxy.prototype.nodeWithIdentifier):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotNodeProxy.js:
(WI.HeapSnapshotNodeProxy):
(WI.HeapSnapshotNodeProxy.deserialize):
(WI.HeapSnapshotNodeProxy.prototype.get target): Added.
(WI.HeapSnapshotNodeProxy.prototype.shortestGCRootPath):
(WI.HeapSnapshotNodeProxy.prototype.dominatedNodes):
(WI.HeapSnapshotNodeProxy.prototype.retainedNodes):
(WI.HeapSnapshotNodeProxy.prototype.retainers):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotProxy.js:
(WI.HeapSnapshotProxy):
(WI.HeapSnapshotProxy.deserialize):
(WI.HeapSnapshotProxy.prototype.get target): Added.
(WI.HeapSnapshotProxy.prototype.get imported):
(WI.HeapSnapshotProxy.prototype.instancesWithClassName):
(WI.HeapSnapshotProxy.prototype.nodeWithIdentifier):
Pass along the `WI.Target` when handling events or creating new proxies.

* Source/WebInspectorUI/UserInterface/Controllers/HeapManager.js:
(WI.HeapManager.prototype.snapshot):
(WI.HeapManager.prototype.getPreview):
(WI.HeapManager.prototype.getRemoteObject):
(WI.HeapManager.prototype.garbageCollected):
Use the relevant `WI.Target` instead of just the main target.

* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotWorkerProxy.js:
(WI.HeapSnapshotWorkerProxy.prototype.createImportedSnapshot):
* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
(HeapSnapshot):
(HeapSnapshot.prototype.updateDeadNodesAndGatherCollectionData):
(HeapSnapshot.prototype.serialize):
(HeapSnapshot.prototype._isNodeGlobalObject):
* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshotWorker.js:
(HeapSnapshotWorker.prototype.createSnapshot):
Pass along a `targetId` so that when new heap snapshots are captured we can determine which previous snapshots to update for what&apos;s still live.
Keep a separate identifier counter for each title instead of a single global one (e.g. &quot;Page 1&quot; and &quot;Worker 1&quot; instead of &quot;Page 1&quot; and &quot;Worker 2&quot;).
Drive-by: Fix an exception after 284791@main caused by `_roots` (and `_labels`) not always being set.

* Source/WebInspectorUI/UserInterface/Models/HeapAllocationsTimelineRecord.js:
(WI.HeapAllocationsTimelineRecord.async fromJSON):
Imported heap snapshots won&apos;t have a target.

* Source/WebInspectorUI/UserInterface/Views/HeapAllocationsTimelineView.js:
(WI.HeapAllocationsTimelineView):
(WI.HeapAllocationsTimelineView.prototype._importButtonNavigationItemClicked):
(WI.HeapAllocationsTimelineView.prototype._takeHeapSnapshotClicked):
(WI.HeapAllocationsTimelineView.prototype._dataGridNodeSelected):
Allow for heap snapshots to be sorted by name since there&apos;s now more than just &quot;Snapshot #&quot;.

* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotClusterContentView.js:
(WI.HeapSnapshotClusterContentView.iconStyleClassNameForClassName):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotDataGridTree.js:
(WI.HeapSnapshotObjectGraphDataGridTree.prototype.populateTopLevel):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode):
(WI.HeapSnapshotInstanceDataGridNode.prototype.createCellContent):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateWindowPreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populatePreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler.appendPath):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler.appendPathRow):
Don&apos;t assume that the root is named &quot;Window&quot;, as `Worker` should be &quot;DedicatedWorkerGlobalScope&quot;.

* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.displayNameForRecord):
Change how heap snapshots are named now that they&apos;re captured in the page and `Worker`.

* Source/WebInspectorUI/UserInterface/Views/TextNavigationItem.css:
(@media (prefers-color-scheme: dark) .navigation-bar .item.text): Added.
Drive-by: Fix heap snapshot diff text color not being visible in dark mode.

* LayoutTests/inspector/heap/getPreview.html:
* LayoutTests/inspector/heap/getRemoteObject.html:
* LayoutTests/inspector/heap/imported-snapshot.html:
* LayoutTests/inspector/heap/snapshot.html:
* LayoutTests/inspector/unit-tests/heap-snapshot.html:
* LayoutTests/inspector/unit-tests/heap-snapshot-collection-event.html:

Canonical link: <a href="https://commits.webkit.org/289847@main">https://commits.webkit.org/289847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ab7cbd6267748837457c74c50f7e5666046e333

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88192 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7709 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42600 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/38942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15889 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/93145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/38942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91194 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/6184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79766 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/5956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/38050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35059 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/94991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15362 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/94991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15618 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75622 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/94991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18730 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20537 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/8389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15380 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/15122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/18568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/16904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -40,6 +40,7 @@ localizedStrings["%d Unsupported (plural)"] = "%d Unsupported";
 localizedStrings["%d Unsupported (singular)"] = "%d Unsupported";
 localizedStrings["%d Warning"] = "%d Warning";
 localizedStrings["%d Warnings"] = "%d Warnings";
+localizedStrings["%d \u2014 \u201C%s\u201D"] = "%d \u2014 \u201C%s\u201D";
```
Please add a label. @_@

burg commented on Source/WebInspectorUI/UserInterface/Base/Setting.js:
```diff
@@ -243,6 +243,7 @@ WI.settings = {
     experimentalShowCaseSensitiveAutocomplete: new WI.Setting("experimental-show-case-sensitive-auto-complete", false),
     experimentalLimitSourceCodeHighlighting: new WI.Setting("engineering-limit-source-code-highlighting", false),
     experimentalUseFuzzyMatchingForCSSCodeCompletion: new WI.Setting("experimental-use-fuzzy-matching-for-css-code-completion", true),
+    experimentalEnableWorkerTimelineRecording: new WI.Setting("experimental-worker-timeline-recording", false),
```
I believe this was already added by the other patch.

dcrousso replied:
i wasnt sure which would land first, so i added it to both 😅

burg commented on Source/WebInspectorUI/UserInterface/Views/HeapSnapshotDataGridTree.js:
```diff
@@ -254,7 +254,7 @@ WI.HeapSnapshotInstancesDataGridTree = class HeapSnapshotInstancesDataGridTree e
     }
 };
 
-WI.HeapSnapshotObjectGraphDataGridTree = class HeapSnapshotInstancesDataGridTree extends WI.HeapSnapshotDataGridTree
```
What??

dcrousso replied:
this was a drive-by as i was having a hard time using "Jump to Symbol" with this because it was named the same as `WI.HeapSnapshotInstancesDataGridTree`

burg commented on Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
```diff
@@ -295,12 +295,17 @@ WI.TimelineTabContentView = class TimelineTabContentView extends WI.ContentBrows
-            return WI.UIString("Snapshot %d").format(timelineRecord.heapSnapshot.identifier);
+        case WI.TimelineRecord.Type.HeapAllocations: {
+            let snapshot = timelineRecord.heapSnapshot;
+
```
Nit: extra newline

burg commented on Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
```diff
@@ -295,12 +295,17 @@ WI.TimelineTabContentView = class TimelineTabContentView extends WI.ContentBrows
+            if (snapshot.imported)
+                return WI.UIString("Imported Snapshot \u2014 %s").format(snapshot.title || snapshot.identifier);
+
+            let title = snapshot.title ? WI.UIString("%d \u2014 \u201C%s\u201D").format(snapshot.identifier, snapshot.title) : snapshot.identifier;
```
@_@

dcrousso committed:
Web Inspector: Enable Memory profiling in Workers
https://bugs.webkit.org/show_bug.cgi?id=167323

Reviewed by BJ Burg.

The Web Inspector frontend always assumes the main target when interacting with `Heap`, meaning that even though `Heap` is already enabled inside `Worker` (i.e. there's even a `WebHeapAgent`) it'll never communicate with any other `WI.Target`.

Note that there are still a few things missing from full support as `Timelines` currently only supports the main target (e.g. auto-capture in `Worker` on page load won't happen).

* Source/WebInspectorUI/UserInterface/Models/HeapAllocationsInstrument.js:
(WI.HeapAllocationsInstrument.prototype.startInstrumentation):
(WI.HeapAllocationsInstrument.prototype.stopInstrumentation):
(WI.HeapAllocationsInstrument.prototype._takeHeapSnapshot):
Start and stop `Heap` for each `WI.Target` instead of just the main target.

* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.heapSnapshot):
* Source/WebInspectorUI/UserInterface/Protocol/HeapObserver.js:
(WI.HeapObserver.prototype.trackingStart):
(WI.HeapObserver.prototype.trackingComplete):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotDiffProxy.js:
(WI.HeapSnapshotDiffProxy):
(WI.HeapSnapshotDiffProxy.deserialize):
(WI.HeapSnapshotDiffProxy.prototype.get target): Added.
(WI.HeapSnapshotDiffProxy.prototype.instancesWithClassName):
(WI.HeapSnapshotDiffProxy.prototype.nodeWithIdentifier):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotNodeProxy.js:
(WI.HeapSnapshotNodeProxy):
(WI.HeapSnapshotNodeProxy.deserialize):
(WI.HeapSnapshotNodeProxy.prototype.get target): Added.
(WI.HeapSnapshotNodeProxy.prototype.shortestGCRootPath):
(WI.HeapSnapshotNodeProxy.prototype.dominatedNodes):
(WI.HeapSnapshotNodeProxy.prototype.retainedNodes):
(WI.HeapSnapshotNodeProxy.prototype.retainers):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotProxy.js:
(WI.HeapSnapshotProxy):
(WI.HeapSnapshotProxy.deserialize):
(WI.HeapSnapshotProxy.prototype.get target): Added.
(WI.HeapSnapshotProxy.prototype.get imported):
(WI.HeapSnapshotProxy.prototype.instancesWithClassName):
(WI.HeapSnapshotProxy.prototype.nodeWithIdentifier):
Pass along the `WI.Target` when handling events or creating new proxies.

* Source/WebInspectorUI/UserInterface/Controllers/HeapManager.js:
(WI.HeapManager.prototype.snapshot):
(WI.HeapManager.prototype.getPreview):
(WI.HeapManager.prototype.getRemoteObject):
(WI.HeapManager.prototype.garbageCollected):
Use the relevant `WI.Target` instead of just the main target.

* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotWorkerProxy.js:
(WI.HeapSnapshotWorkerProxy.prototype.createImportedSnapshot):
* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
(HeapSnapshot):
(HeapSnapshot.prototype.updateDeadNodesAndGatherCollectionData):
(HeapSnapshot.prototype.serialize):
(HeapSnapshot.prototype._isNodeGlobalObject):
* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshotWorker.js:
(HeapSnapshotWorker.prototype.createSnapshot):
Pass along a `targetId` so that when new heap snapshots are captured we can determine which previous snapshots to update for what's still live.
Keep a separate identifier counter for each title instead of a single global one (e.g. "Page 1" and "Worker 1" instead of "Page 1" and "Worker 2").
Drive-by: Fix an exception after 284791@main caused by `_roots` (and `_labels`) not always being set.

* Source/WebInspectorUI/UserInterface/Models/HeapAllocationsTimelineRecord.js:
(WI.HeapAllocationsTimelineRecord.async fromJSON):
Imported heap snapshots won't have a target.

* Source/WebInspectorUI/UserInterface/Views/HeapAllocationsTimelineView.js:
(WI.HeapAllocationsTimelineView):
(WI.HeapAllocationsTimelineView.prototype._importButtonNavigationItemClicked):
(WI.HeapAllocationsTimelineView.prototype._takeHeapSnapshotClicked):
(WI.HeapAllocationsTimelineView.prototype._dataGridNodeSelected):
Allow for heap snapshots to be sorted by name since there's now more than just "Snapshot #".

* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotClusterContentView.js:
(WI.HeapSnapshotClusterContentView.iconStyleClassNameForClassName):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotDataGridTree.js:
(WI.HeapSnapshotObjectGraphDataGridTree.prototype.populateTopLevel):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode):
(WI.HeapSnapshotInstanceDataGridNode.prototype.createCellContent):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateWindowPreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populatePreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler.appendPath):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler.appendPathRow):
Don't assume that the root is named "Window", as `Worker` should be "DedicatedWorkerGlobalScope".

* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.displayNameForRecord):
Change how heap snapshots are named now that they're captured in the page and `Worker`.

* Source/WebInspectorUI/UserInterface/Views/TextNavigationItem.css:
(@media (prefers-color-scheme: dark) .navigation-bar .item.text): Added.
Drive-by: Fix heap snapshot diff text color not being visible in dark mode.

* LayoutTests/inspector/heap/getPreview.html:
* LayoutTests/inspector/heap/getRemoteObject.html:
* LayoutTests/inspector/heap/imported-snapshot.html:
* LayoutTests/inspector/heap/snapshot.html:
* LayoutTests/inspector/unit-tests/heap-snapshot.html:
* LayoutTests/inspector/unit-tests/heap-snapshot-collection-event.html:

Canonical link: https://commits.webkit.org/289847@main
